experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeBinder.ts

```diff
@@ -0,0 +1,1003 @@
+		new Map();
+	public constructor(protected readonly options: BinderOptions) {}
+	public abstract onDelete(path: UpPath, count: number): void;
+	public abstract onInsert(path: UpPath, content: ProtoNodes): void;
+	public abstract onSetValue(path: UpPath, value: TreeValue): void;
+	public registerListener(
+		contextType: BindingContextType,
+		trees: BindTree[],
+		listener: Listener,
+	): () => void {
+		let contextRoots = this.registeredListeners.get(contextType);
+		if (contextRoots === undefined) {
+			contextRoots = new Map();
+			this.registeredListeners.set(contextType, contextRoots);
+		}
+		trees.forEach((tree) => {
+			const currentRoot = this.findRoot(contextType, tree.field);
+			if (currentRoot === undefined) {
+				const newRoot: CallTree = {
+					field: tree.field,
+					index: tree.index,
+					listeners: new Set(),
+					children: new Map(),
+				};
+				assert(contextRoots !== undefined, "contextRoots are defined");
+				contextRoots.set(tree.field, newRoot);
+				this.bindTree(contextType, tree, listener, newRoot);
+			} else {
+				this.bindTree(contextType, tree, listener, currentRoot);
+			}
+		});
+		return () => {
+			trees.forEach((tree) => this.unregisterListener(contextType, tree, listener));
+		};
+	}
+
+	private bindTree(
+		contextType: BindingContextType,
+		tree: BindTree,
+		listener: Listener,
+		callTree: CallTree,
+	) {
+		if (tree.children.size === 0) {
+			callTree.listeners.add(listener);
+		} else {
+			tree.children.forEach((childTree, fieldKey) => {
+				let childCallTree: CallTree | undefined = callTree.children.get(fieldKey);
+				if (childCallTree === undefined) {
+					childCallTree = {
+						field: fieldKey,
+						index: childTree.index,
+						listeners: new Set(),
+						children: new Map(),
+					};
+					callTree.children.set(fieldKey, childCallTree);
+				}
+				this.bindTree(contextType, childTree, listener, childCallTree);
+			});
+		}
+	}
+
+	private findRoot(contextType: BindingContextType, field: FieldKey): CallTree | undefined {
+		return this.registeredListeners.get(contextType)?.get(field) ?? undefined;
+	}
+
+	private unregisterListener(
+		contextType: BindingContextType,
+		tree: BindTree,
+		listener: Listener,
+		callTree?: CallTree,
+	) {
+		let foundTree = callTree;
+		if (foundTree === undefined) {
+			foundTree = this.findRoot(contextType, tree.field);
+		}
+		if (foundTree !== undefined) {
+			if (tree.children.size === 0) {
+				foundTree.listeners.delete(listener);
+			} else {
+				tree.children.forEach((childTree, fieldKey) => {
+					assert(foundTree !== undefined, "foundTree is not undefined");
+					const childCallTree = foundTree.children.get(fieldKey);
+					if (childCallTree) {
+						this.unregisterListener(contextType, childTree, listener, childCallTree);
+					}
+				});
+			}
+		}
+	}
+
+	protected getListeners(
+		contextType: BindingContextType,
+		downPath: DownPath,
+	): Set<Listener> | undefined {
```

Comment:
I justify `undefined` vs. a uniform return (1) to avoid the overhead of set instantiation considering also the high-frequency of the notification calls, (2) support consumers which need to make an inexpensive decision whether content is available or not w/o consulting the size property (which should be typically more expensive) and (3) the fact that consumers are limited to `DataBinder` itself, ie internal.  If acceptable I would like to preserve the pattern but still open to suggestions.

Reply:
Those are good reasons, thanks! 

(1) can be solved by creating a singleton `const emptySet: ReadonlySet<Listener> = new Set()` at the top of the file and reusing it for the empty case (and making the return type of `getListeners` a `ReadonlySet`; that is probably a good thing to do regardless). 

I'm not sure about (2), although I'm skeptical that reading `size` is meaningfully slower than checking `undefined`. I think it is just stored as a number property on the set object, so it should always be O(1) read:
![image](https://github.com/microsoft/FluidFramework/assets/78610362/050784ea-8c48-42e1-91ae-e554d688be71)
Well, unless we later implement a custom kind of `Set` with an O(n) `size`. I suppose that is possible 😅 .

But, (3) is a great reason to keep it the same. If you, the author, prefer the pattern as it is, then it's fine with me :)